test/e2e/framework/test.go

```diff
@@ -0,0 +1,100 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+
+	"k8s.io/client-go/rest"
+)
+
+// we want to ensure that everything runs in parallel and that users of this framework
+// do not need to explicitly remember to call the top-level testing.T method, so we do
+// it for them and keep track of which we've already done so we do not cause a double-
+// call panic
+var seen = sync.Map{}
+
+type RunningServer interface {
+	Config() *rest.Config
+}
+
+type TestFunc func(t TestingTInterface, servers ...RunningServer)
+
+// Run mimics the testing.T.Run function while providing a nice set of concurrency
+// guarantees for the processes that we create and manage for test cases. We ensure:
+// - any kcp processes that are started will only be exposed to the test
+//   code once they have signalled that they are healthy, ready, and live
+// - any kcp processes will have their lifetime bound to the lifetime of the
+//   individual test case - they will be cancelled when the test finishes
+// - any errors in running an accessory process (other than it being cancelled by the
+//   above mechanism) will be fatal to the test execution and will preempt the other
+//   test routines, including failures to continue being healthy, ready and live
+// This is a generally non-standard use of the testing.T construct; the two largest
+// reasons we built this abstraction are that we need to manage a fairly complex and
+// fragile set of concurrency and lifetime guarantees for the processes that are
+// executed for each test case. We must control the scope in which each test runs, to
+// be able to defer() actions that cancel background processes when a test finishes
+// execution, as child goroutines are not by default killed when a test finishes and
+// interaction with the parent testing.T after the test is finished causes a panic.
+// Furthermore, we could not simply use the central testing.T for managing control
+// flow as it is not allowed to call testing.T.FailNow (via Fatal, etc) in anything
+// other than the main testing goroutine. Therefore, when more than one routine needs
+// to be able to influence the execution flow (e.g. preempt other routines) we must
+// have the central routine watch for incoming errors from delegate routines.
+func Run(top *testing.T, name string, f TestFunc, args ...[]string) {
+	if _, previouslyCalled := seen.LoadOrStore(fmt.Sprintf("%p", top), nil); !previouslyCalled {
```

Comment:
When would `top` be seen multiple times - multiple Test funcs in the same package that call `framework.Run`?

Reply:
Yep, exactly in that case